File: apps/macos/FlixorMac/Services/MDBList/MDBListService.swift
//
//  MDBListService.swift
//  FlixorMac
//
//  MDBList API service for fetching ratings from multiple sources
//  Requires user to enable and provide API key in settings
//

import Foundation

// MARK: - API Response Models

private struct MDBListRatingResponse: Codable {
    let ratings: [MDBListRatingItem]?
}

private struct MDBListRatingItem: Codable {
    let id: String?
    let rating: Double?
}

// MARK: - MDBListService

@MainActor
class MDBListService: ObservableObject {
    // MARK: - Singleton

    static let shared = MDBListService()

    // MARK: - Configuration

    private let baseURL = "https://api.mdblist.com"
    private let cacheTTL: TimeInterval = 24 * 60 * 60 // 24 hours

    // MARK: - Cache

    private var cache: [String: (ratings: MDBListRatings?, timestamp: Date)] = [:]

    // MARK: - Computed Properties

    private var isEnabled: Bool {
        UserDefaults.standard.bool(forKey: "mdblistEnabled")
    }

    private var apiKey: String? {
        let key = UserDefaults.standard.string(forKey: "mdblistApiKey")
        return key?.isEmpty == false ? key : nil
    }

    // MARK: - Initialization

    private init() {}

    // MARK: - Public API

    /// Check if MDBList is enabled and has an API key
    func isReady() -> Bool {
        return isEnabled && apiKey != nil
    }

    /// Fetch ratings from MDBList API
    /// - Parameters:
    ///   - imdbId: The IMDb ID (with or without "tt" prefix)
    ///   - mediaType: Either "movie" or "show"
    /// - Returns: MDBListRatings or nil if unavailable
    func fetchRatings(imdbId: String, mediaType: String) async -> MDBListRatings? {
        guard isEnabled else {
            print("[MDBListService] MDBList is disabled")
            return nil
        }

        guard let apiKey = apiKey else {
            print("[MDBListService] No API key configured")
            return nil
        }

        // Normalize IMDb ID
        let formattedImdbId = imdbId.hasPrefix("tt") ? imdbId : "tt\(imdbId)"
        guard formattedImdbId.range(of: "^tt\\d+$", options: .regularExpression) != nil else {
            print("[MDBListService] Invalid IMDb ID format: \(formattedImdbId)")
            return nil
        }

        // Check cache
        let cacheKey = "\(mediaType):\(formattedImdbId)"
        if let cached = cache[cacheKey],
           Date().timeIntervalSince(cached.timestamp) < cacheTTL {
            print("[MDBListService] Cache hit for \(cacheKey): \(cached.ratings != nil ? "found" : "null")")
            return cached.ratings
        }

        print("[MDBListService] Fetching ratings for \(mediaType): \(formattedImdbId)")

        var ratings = MDBListRatings()
        let ratingTypes = ["imdb", "tmdb", "trakt", "letterboxd", "tomatoes", "audience", "metacritic"]

        // Fetch all rating types in parallel
        await withTaskGroup(of: (String, Double?).self) { group in
            for ratingType in ratingTypes {
                group.addTask {
                    await self.fetchSingleRating(
                        mediaType: mediaType,
                        ratingType: ratingType,
                        imdbId: formattedImdbId,
                        apiKey: apiKey
                    )
                }
            }

            for await (type, rating) in group {
                switch type {
                case "imdb": ratings.imdb = rating
                case "tmdb": ratings.tmdb = rating
                case "trakt": ratings.trakt = rating
                case "letterboxd": ratings.letterboxd = rating
                case "tomatoes": ratings.tomatoes = rating
                case "audience": ratings.audience = rating
                case "metacritic": ratings.metacritic = rating
                default: break
                }
            }
        }

        let hasAnyRating = ratings.hasAnyRating
        print("[MDBListService] Fetched ratings - hasAny: \(hasAnyRating)")

        // Cache the result (even if empty to avoid repeated failed requests)
        let finalResult = hasAnyRating ? ratings : nil
        cache[cacheKey] = (ratings: finalResult, timestamp: Date())

        return finalResult
    }

    /// Clear the ratings cache
    func clearCache() {
        cache.removeAll()
        print("[MDBListService] Cache cleared")
    }

    // MARK: - Private Helpers

    private func fetchSingleRating(
        mediaType: String,
        ratingType: String,
        imdbId: String,
        apiKey: String
    ) async -> (String, Double?) {
        let urlString = "\(baseURL)/rating/\(mediaType)/\(ratingType)?apikey=\(apiKey)"

        guard let url = URL(string: urlString) else {
            return (ratingType, nil)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "ids": [imdbId],
            "provider": "imdb"
        ]

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)

            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                return (ratingType, nil)
            }

            if httpResponse.statusCode == 403 {
                print("[MDBListService] Invalid API key")
                return (ratingType, nil)
            }

            if httpResponse.statusCode == 200 {
                let decoded = try JSONDecoder().decode(MDBListRatingResponse.self, from: data)
                if let rating = decoded.ratings?.first?.rating {
                    return (ratingType, rating)
                }
            }
        } catch {
            print("[MDBListService] Error fetching \(ratingType): \(error)")
        }

        return (ratingType, nil)
    }
}

File: apps/mobile/src/core/MDBListService.ts
/**
 * MDBList API Service for fetching ratings from multiple sources
 * Requires user to enable and provide API key in settings
 */

import { isMdblistEnabled, getMdblistApiKey } from './SettingsData';

const MDBLIST_BASE_URL = 'https://api.mdblist.com';
const CACHE_TTL = 24 * 60 * 60 * 1000; // 24 hours

export interface MDBListRatings {
  imdb?: number;
  tmdb?: number;
  trakt?: number;
  letterboxd?: number;
  tomatoes?: number; // Rotten Tomatoes critic score
  audience?: number; // Rotten Tomatoes audience score
  metacritic?: number;
}

// In-memory cache for ratings
const ratingsCache = new Map<string, { ratings: MDBListRatings | null; timestamp: number }>();

/**
 * Check if MDBList is enabled and has an API key
 */
export function isMDBListReady(): boolean {
  return isMdblistEnabled() && !!getMdblistApiKey();
}

/**
 * Fetch ratings from MDBList API
 */
export async function fetchMDBListRatings(
  imdbId: string,
  mediaType: 'movie' | 'show'
): Promise<MDBListRatings | null> {
  // Check if MDBList is enabled
  if (!isMdblistEnabled()) {
    console.log('[MDBListService] MDBList is disabled');
    return null;
  }

  const apiKey = getMdblistApiKey();
  if (!apiKey) {
    console.log('[MDBListService] No API key configured');
    return null;
  }

  // Normalize IMDb ID
  const formattedImdbId = imdbId.startsWith('tt') ? imdbId : `tt${imdbId}`;
  if (!/^tt\d+$/.test(formattedImdbId)) {
    console.log('[MDBListService] Invalid IMDb ID format:', formattedImdbId);
    return null;
  }

  // Check cache
  const cacheKey = `${mediaType}:${formattedImdbId}`;
  const cached = ratingsCache.get(cacheKey);
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    console.log(`[MDBListService] Cache hit for ${cacheKey}:`, cached.ratings ? 'found' : 'null');
    return cached.ratings;
  }

  try {
    console.log(`[MDBListService] Fetching ratings for ${mediaType}: ${formattedImdbId}`);

    const ratings: MDBListRatings = {};
    const ratingTypes = ['imdb', 'tmdb', 'trakt', 'letterboxd', 'tomatoes', 'audience', 'metacritic'];

    // Fetch all rating types in parallel
    const fetchPromises = ratingTypes.map(async (ratingType) => {
      try {
        const url = `${MDBLIST_BASE_URL}/rating/${mediaType}/${ratingType}?apikey=${apiKey}`;

        const response = await fetch(url, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            ids: [formattedImdbId],
            provider: 'imdb'
          })
        });

        if (response.ok) {
          const data = await response.json();
          if (data.ratings?.[0]?.rating) {
            return { type: ratingType, rating: data.ratings[0].rating };
          }
        } else if (response.status === 403) {
          console.log('[MDBListService] Invalid API key');
        }
        return null;
      } catch (error) {
        console.log(`[MDBListService] Error fetching ${ratingType}:`, error);
        return null;
      }
    });

    const results = await Promise.all(fetchPromises);

    results.forEach(result => {
      if (result) {
        ratings[result.type as keyof MDBListRatings] = result.rating;
      }
    });

    const ratingCount = Object.keys(ratings).length;
    console.log(`[MDBListService] Fetched ${ratingCount} ratings:`, ratings);

    // Cache the result (even if empty to avoid repeated failed requests)
    const finalResult = ratingCount > 0 ? ratings : null;
    ratingsCache.set(cacheKey, { ratings: finalResult, timestamp: Date.now() });

    return finalResult;
  } catch (error) {
    console.log('[MDBListService] Error fetching ratings:', error);
    return null;
  }
}

/**
 * Clear the ratings cache
 */
export function clearMDBListCache(): void {
  ratingsCache.clear();
  console.log('[MDBListService] Cache cleared');
}

/**
 * Rating provider configuration for UI display
 */
export const RATING_PROVIDERS = {
  imdb: {
    name: 'IMDb',
    color: '#F5C518',
  },
  tmdb: {
    name: 'TMDB',
    color: '#01B4E4',
  },
  trakt: {
    name: 'Trakt',
    color: '#ED1C24',
  },
  letterboxd: {
    name: 'Letterboxd',
    color: '#00E054',
  },
  tomatoes: {
    name: 'Rotten Tomatoes',
    color: '#FA320A',
  },
  audience: {
    name: 'Audience Score',
    color: '#FA320A',
  },
  metacritic: {
    name: 'Metacritic',
    color: '#FFCC33',
  }
} as const;
